### ISSUE TITLE
Limit JNI calls in nativePixelForLatLng & nativeLatLngForPixel

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#5104]
In `jni.cpp` we have 2 methods that are called often:
 - nativePixelForLatLng
 - nativeLatLngForPixel

These methods are used to translate a pixel on the screen to a certain location coordinate and reversed. 

Currently these methods look like:

```cpp
jni::jobject* nativePixelForLatLng(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jni::jobject* latLng) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativePixelForLatLng");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);

    jdouble latitude = jni::GetField<jdouble>(*env, latLng, *latLngLatitudeId);
    jdouble longitude = jni::GetField<jdouble>(*env, latLng, *latLngLongitudeId);

    mbgl::ScreenCoordinate pixel = nativeMapView->getMap().pixelForLatLng(mbgl::LatLng(latitude, longitude));
    return &jni::NewObject(*env, *pointFClass, *pointFConstructorId, static_cast<jfloat>(pixel.x), static_cast<jfloat>(pixel.y));
}

jni::jobject* nativeLatLngForPixel(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jni::jobject* pixel) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativeLatLngForPixel");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);

    jfloat x = jni::GetField<jfloat>(*env, pixel, *pointFXId);
    jfloat y = jni::GetField<jfloat>(*env, pixel, *pointFYId);

    mbgl::LatLng latLng = nativeMapView->getMap().latLngForPixel(mbgl::ScreenCoordinate(x, y));
    return &jni::NewObject(*env, *latLngClass, *latLngConstructorId, latLng.latitude, latLng.longitude);
}
```
Each JNI method execution above results in 4 JNI rounds trips:
 - JNI method itself
 -  jni::GetField
 -  jni::GetField
 -  jni::NewObject

Following the [guidelines](https://developer.android.com/training/articles/perf-jni.html) related to JNI and mentioned in the [Google I/O session related to NDK](https://www.youtube.com/watch?v=Ok97X9Z4Si8), we should limit the amount of JNI calls inside one JNI method. In case from above this would result in removing the `jni::GetField`/ `jni::NewObject` items and replacing `jobject pixel` and `jobject latlng` with multiple parameters.

**Comment by tobrun** [comment]
I started looking into profiling small changes:

```cpp
jni::jobject* nativePixelForLatLng(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jni::jobject* latLng) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativePixelForLatLng");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);

    jdouble latitude = jni::GetField<jdouble>(*env, latLng, *latLngLatitudeId);
    jdouble longitude = jni::GetField<jdouble>(*env, latLng, *latLngLongitudeId);

    mbgl::ScreenCoordinate pixel = nativeMapView->getMap().pixelForLatLng(mbgl::LatLng(latitude, longitude));
    return &jni::NewObject(*env, *pointFClass, *pointFConstructorId, static_cast<jfloat>(pixel.x), static_cast<jfloat>(pixel.y));
}

jni::jobject* nativePixelForLatLng2(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jdouble latitude, jdouble longitude) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativePixelForLatLng2");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);

    mbgl::ScreenCoordinate pixel = nativeMapView->getMap().pixelForLatLng(mbgl::LatLng(latitude, longitude));
    return &jni::NewObject(*env, *pointFClass, *pointFConstructorId, static_cast<jfloat>(pixel.x), static_cast<jfloat>(pixel.y));
}
```

Profiling execution time with 1000 calls:
 - nativePixelForLatLng: **74ms**
 - nativePixelForLatLng2: **57ms**

**Comment by tobrun** [comment]
##### Remove JNI calls to get Pixel values:

```cpp
jni::jobject* nativeLatLngForPixel(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jni::jobject* pixel) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativeLatLngForPixel");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);

    jfloat x = jni::GetField<jfloat>(*env, pixel, *pointFXId);
    jfloat y = jni::GetField<jfloat>(*env, pixel, *pointFYId);

    mbgl::LatLng latLng = nativeMapView->getMap().latLngForPixel(mbgl::ScreenCoordinate(x, y));
    return &jni::NewObject(*env, *latLngClass, *latLngConstructorId, latLng.latitude, latLng.longitude);
}

jni::jobject* nativeLatLngForPixel2(JNIEnv *env, jni::jobject* obj, jlong nativeMapViewPtr, jfloat x, jfloat y) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativeLatLngForPixel");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);
    mbgl::LatLng latLng = nativeMapView->getMap().latLngForPixel(mbgl::ScreenCoordinate(x, y));
    return &jni::NewObject(*env, *latLngClass, *latLngConstructorId, latLng.latitude, latLng.longitude);
}
```

Profiling execution time with 1000 calls:
- nativeLatLngForPixel1: **52ms**
- nativeLatLngForPixel2: **45ms**

**Comment by tobrun** [comment]
I'm now looking into the possibility to used a cached field setup for high trafficking jni calls and see if that is faster.

**Comment by tobrun** [comment]
Closed with #5412 and squashed into 		2bf4e20
